prdoc/pr_4102.prdoc

```diff
@@ -0,0 +1,43 @@
+title: "Bridge: make some headers submissions free"
+
+doc:
+  - audience: Runtime Dev
+    description: |
+      Adds `FreeHeadersInterval` configuration constant to the `pallet_bridge_grandpa`.
+      Transactions that improve best known header by at least `FreeHeadersInterval` headers
+      are now free for the submitter. Additionally, we allow single free parachain header
+      update per every free relay chain header. Bridge signed extensions are adjusted
```

Comment:
I think a malicious relayer can do high-tip free txs that advance the view of the bridged relay chain by `FreeHeadersInterval`, but only advance some parachain view by 1 block, effectively DoS-ing the free update of the parachain.
Or the updated parachain head have to match the `para_heads` of the imported relay block proof?

Even so, attacker(s) would be spending money (the tx tips) for keeping the attack ongoing, while any honest relayer can still update the parachain view (but not for free).

Reply:
> Or the updated parachain head have to match the para_heads of the imported relay block proof?

Maybe I'b bad in describing behavior. Parachain head submission is free (`Pays::No`) iff:
1) best parachain head, known to the pallet is `PH_old` and it is proved at relay chain header `RH_old`;
2) new parachain head `PH_new` is proved at relay chain head `RH_new`;
3) the `RH_new - RH_old` is larger than the `FreeHeadersInterval`.

So no, attacker will be able to increase parachain head number by one for free only if it has actually been increased by one during `FreeHeadersInterval` relay chain headers.

We also have this new signed extension that boosts priority of relay chain header/parachain head submission transactions based on that `RH_new - RH_old` expression. So transaction of registered relayer that submits free header `1000` will have larger priority than transaction of registered relayer that submits free header `900`. So if there's a honest relayer, which tries to submit best available rela/para header `1000`, it will be costly for attacker to submit headers `900`, `910`, `920`, ... for free (assuming `FreeHeadersInterval = 10`) - honest collators will include transaction with `1000`  before other transactions